src/internet_identity/src/state/temp_keys.rs

```diff
@@ -25,56 +39,98 @@ pub struct TempKeys {
     ///
     /// Since temp keys can only be added during registration, the the max number of temp keys is
     /// bounded by the registration rate limit.
-    temp_keys: HashMap<DeviceKey, TempKey>,
+    device_bound_keys: HashMap<DeviceKey, DeviceBoundTempKey>,
 
     /// Deque to efficiently prune expired temp keys
     expirations: VecDeque<TempKeyExpiration>,
 }
 
 impl TempKeys {
-    pub fn add_temp_key(
+    pub fn add_onboarding_temp_key(&mut self, temp_key: Principal) -> Result<(), ()> {
+        self.prune_expired_keys();
+
+        // tie the max number of onboarding keys to the registration rate limit
+        if self.onboarding_keys.len() as u64
+            > state::persistent_state(|ps| {
+                ps.registration_rate_limit
+                    .as_ref()
+                    .map(|rate_limit| rate_limit.max_tokens)
+                    .unwrap_or(20_000)
+            })
+        {
+            return Err(());
+        }
+
+        let expiration = time() + TEMP_KEY_EXPIRATION_NS;
+        self.expirations.push_back(TempKeyExpiration::Onboarding {
+            principal: temp_key,
+            expiration,
+        });
+        self.onboarding_keys.insert(temp_key, expiration);
+        Ok(())
+    }
+
+    pub fn add_device_bound_temp_key(
         &mut self,
         device_key: &DeviceKey,
         anchor: AnchorNumber,
         temp_key: Principal,
     ) {
-        let tmp_key = TempKey {
+        // Remove the temp key from onboarding keys if it exists
+        self.onboarding_keys.remove(&temp_key);
```

Comment:
Is it a bug if the key exists? Or maybe will be a bug if the key does not exist once we fully deprecate the API?

Reply:
No. Later on, it will become a bug. Each temp key will need to have submitted a captcha solution (i.e. be in the list of onboarding keys) to be transferred to a device bound key.